### PR TITLE
Group common Qblox modules functions

### PR DESCRIPTION
**Pull request by PiergiorgioButtarini**
This PR follows the philosophy established in https://github.com/qiboteam/qibolab/pull/686 where common properties between the three diferent Qblox modules are defined directly in the class `ClusterModule` (in `module.py`) and then are inherited by all of them.
 
#### TODO:
Test on hardware.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

**Comment by alecandido** [comment]
This is also outdated wrt `main`. If you know that is working, and the review is just the routine check, please merge/rebase.

Otherwise, make sure that the branch is updated, and you'd like to merge (I'm asking just because it's not very recent).

**Comment by stavros11** [comment]
@alecandido I put as as reviewers because I think in the last Qibo meeting @PiergiorgioButtarini said that this is ready to merge and overal is reducing the lines of code so would be good to merge. But indeed let's wait for him to fix the conflict and let us know.

**Comment by alecandido** [comment]
Yes, I got your point, but there are a few PRs that are aging a bit in Qibolab, so we should be slightly more careful when merging.

File: src/qibolab/instruments/qblox/cluster_qcm_bb.py [hunk 1]
```diff
@@ -160,29 +146,24 @@ def connect(self, cluster: QbloxCluster = None):
...
+            for port in self._ports:
+                self._sequencers[port] = []
+                self._ports[port].hardware_mod_en = True
+                self._ports[port].nco_freq = 0
+                self._ports[port].nco_phase_offs = 0
+        except Exception as error:
+            raise RuntimeError(
+                f"Unable to initialize port parameters on module {self.name}: {error}"
+            )
+        self.is_connected = True
```
Comment (alecandido) [inline src/qibolab/instruments/qblox/cluster_qcm_bb.py:166]: This function is quite repeated as is in the other modules.

I.e. the part that is written here (`QcmBb`) is almost completely repeated, while they may have extra bits inside the `try:` block.

I'd suggest lifting the `connect()` function to the base class, and leave to each subclass just the part inside the `try:` (just make an abstractmethod in the base class, `_setup_ports()` or something like, and implement in the subclasses).

File: src/qibolab/instruments/qblox/cluster_qrm_rf.py [hunk 2]
```diff
@@ -276,22 +260,9 @@ def _get_next_sequencer(self, port: str, frequency: int, qubits: dict, qubit: No
...
-                ]:
-                    value = self.device.sequencers[self.DEFAULT_SEQUENCERS[port]].get(
-                        param_name=parameter
-                    )
-                    if value:
-                        target = self.device.sequencers[next_sequencer_number]
-                        target.set(parameter, value)
+        default_sequencer_number = self.DEFAULT_SEQUENCERS[port]
+        if next_sequencer_number != default_sequencer_number:
+            self.clone_sequencer_params(default_sequencer_number, next_sequencer_number)
```
Comment (alecandido) [inline src/qibolab/instruments/qblox/cluster_qrm_rf.py:265]: Also, this whole pattern is repeated. In principle, you could split even this.

**Review by alecandido — APPROVED** [review]
Just a few suggestions for further improvements.

But it's already deduplicating a lot, so it's also fine to merge :)

File: src/qibolab/instruments/qblox/cluster_qcm_bb.py [hunk 3]
```diff
@@ -114,19 +109,10 @@ def __init__(self, name: str, address: str):
         >>> qcm_module = QcmBb(name="qcm_bb", address="192.168.1.100:2", cluster=cluster_instance)
         """
         super().__init__(name, address)
-        self._ports: dict = {}
-        self.device: QbloxQrmQcm = None
 
         self._debug_folder: str = ""
         self._sequencers: dict[Sequencer] = {}
-        self.channel_map: dict = {}
         self._device_num_output_ports = 2
```
[the remark this answers is not in the document]
Comment (PiergiorgioButtarini) [reply src/qibolab/instruments/qblox/cluster_qcm_bb.py:115]: Actually `self._device_num_output_ports` is deprecated so I will remove it!

File: src/qibolab/instruments/qblox/cluster_qcm_bb.py (hunk 1 above)
Comment (PiergiorgioButtarini) [reply src/qibolab/instruments/qblox/cluster_qcm_bb.py:166]: Thanks for the suggestion!

File: src/qibolab/instruments/qblox/cluster_qrm_rf.py (hunk 2 above)
Comment (PiergiorgioButtarini) [reply src/qibolab/instruments/qblox/cluster_qrm_rf.py:265]: On this I prefer to work when I will able to test on hardware. Also I'm really tempted to rewrite all the `process_pulse_sequence()` so idk if it's worth making it look better.

File: src/qibolab/instruments/qblox/cluster_qcm_bb.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (PiergiorgioButtarini) [reply src/qibolab/instruments/qblox/cluster_qcm_bb.py]: Solved in [0004e88](https://github.com/qiboteam/qibolab/pull/730/commits/0004e8855ec34d876d7303e6b53e4b55809f4654).

File: src/qibolab/instruments/qblox/cluster_qcm_bb.py [hunk 4]
```diff
@@ -114,19 +109,10 @@ def __init__(self, name: str, address: str):
         >>> qcm_module = QcmBb(name="qcm_bb", address="192.168.1.100:2", cluster=cluster_instance)
         """
         super().__init__(name, address)
-        self._ports: dict = {}
-        self.device: QbloxQrmQcm = None
 
         self._debug_folder: str = ""
         self._sequencers: dict[Sequencer] = {}
```
[the remark this answers is not in the document]
Comment (PiergiorgioButtarini) [reply src/qibolab/instruments/qblox/cluster_qcm_bb.py]: Solved in [0004e88](https://github.com/qiboteam/qibolab/pull/730/commits/0004e8855ec34d876d7303e6b53e4b55809f4654).

File: src/qibolab/instruments/qblox/cluster_qcm_bb.py (hunk 1 above)
Comment (PiergiorgioButtarini) [reply src/qibolab/instruments/qblox/cluster_qcm_bb.py]: Solved [0004e88](https://github.com/qiboteam/qibolab/pull/730/commits/0004e8855ec34d876d7303e6b53e4b55809f4654).

**Comment by alecandido** [comment]
Pylint is failing
https://github.com/qiboteam/qibolab/actions/runs/7844876458/job/21408077607?pr=730#step:10:17

**Review by stavros11 — APPROVED** [review]
Thanks @PiergiorgioButtarini. I am not sure if there is any qblox available right now and where is the runcard, but if you have access it would be good if you can check that at least instrument tests pass before merging.

File: src/qibolab/instruments/qblox/cluster_qcm_rf.py [hunk 5]
```diff
@@ -165,46 +150,16 @@ def _set_default_values(self):
...
+    def _setup_ports(self):
+        for port in self.settings:
+            self._sequencers[port] = []
+            if self.settings[port]["lo_frequency"]:
+                self._ports[port].lo_enabled = True
+                self._ports[port].lo_frequency = self.settings[port]["lo_frequency"]
+            self._ports[port].attenuation = self.settings[port]["attenuation"]
+            self._ports[port].hardware_mod_en = True
+            self._ports[port].nco_freq = 0
+            self._ports[port].nco_phase_offs = 0
```
[the remark this answers is not in the document]
Comment (PiergiorgioButtarini) [reply src/qibolab/instruments/qblox/cluster_qcm_rf.py:162]: Yes you are probably right, I will have a look. Thanks @stavros11

**Comment by alecandido** [comment]
Superseded by #868 and the subsequent rewriting

**Pull request closed without merging**